blocks/graph.py

```diff
@@ -196,8 +197,14 @@ def replace(self, replacements):
         # Due to theano specifics we have to make one replacement in time
         replacements = OrderedDict(replacements)
 
-        apply_nodes_sorted = theano.gof.graph.io_toposort(
-            self.inputs, self.outputs)
+        for variable in replacements:
+            if variable not in self.variables:
+                warnings.warn("replace method was asked to replace a variable "
+                              "that is not a part of the computational graph.")
+            if variable in self.auxiliary_variables:
```

Comment:
has_roles(variable, [AUXILIARY]) will avoid searching in a list. Change if to elif to skip variables that are not part of the graph (we don't care if it's auxiliary, it's not in the graph).